bftengine/tests/clientsManager/ClientsManager_test.cpp

```diff
@@ -32,13 +249,1153 @@ TEST(ClientsManager, reservedPagesPerClient) {
+                "This unit test for ClientsManager::removeRequestsOutOfBatchBounds assumes a minimum value for the "
+                "global system constant maxNumOfRequestsInBatch (which influences removeRequestsOutOfBatchBounds's "
+                "behavior), but that assumption has not been met.");
+
+  unique_ptr<ClientsManager> cm(new ClientsManager({8, 12}, {4, 5, 7}, {10, 11}, metrics));
+  for (uint32_t i = 1; i < maxNumOfRequestsInBatch; ++i) {
+    cm->addPendingRequest(8, i, "correlation ID");
+  }
+  cm->addPendingRequest(8, maxNumOfRequestsInBatch + 1, "correlation ID");
+
+  cm->removeRequestsOutOfBatchBounds(8, maxNumOfRequestsInBatch);
+  EXPECT_FALSE(cm->isClientRequestInProcess(8, maxNumOfRequestsInBatch + 1))
+      << "ClientsManager::removeRequestOutOfBatchBounds failed to remove the request record with the greatest sequence "
+         "number when that sequence number exceeded the one passed to it as a parameter, that sequence number passed "
+         "as a parameter did not match any request records the ClientsManager should have had, and when the "
+         "ClientsManager should have had maxNumOfRequestsInBatch request records.";
+  for (uint32_t i = 1; i < maxNumOfRequestsInBatch; ++i) {
+    EXPECT_TRUE(cm->isClientRequestInProcess(8, i))
+        << "ClientsManager::removeRequestsOutOfBatchBounds removed a request record other than the one with the "
+           "greatest sequence number.";
+  }
+
+  cm->addPendingRequest(8, maxNumOfRequestsInBatch, "correlation ID");
+  cm->removeRequestsOutOfBatchBounds(8, maxNumOfRequestsInBatch + 1);
+  EXPECT_TRUE(cm->isClientRequestInProcess(8, maxNumOfRequestsInBatch))
+      << "ClientsManager::removeRequestsOutOfBatchBounds removed the request record with the greatest sequence number "
+         "in spite of that sequence number being less than the one passed to it as a parameter.";
+  cm->removeRequestsOutOfBatchBounds(8, maxNumOfRequestsInBatch);
+  EXPECT_TRUE(cm->isClientRequestInProcess(8, maxNumOfRequestsInBatch))
+      << "ClientsManager::removeRequestsOutOfBatchBounds removed the request record with the greatest sequence number "
+         "in spite of that sequence number equalling the one passed to it as a parameter.";
+
+  cm->removePendingForExecutionRequest(8, maxNumOfRequestsInBatch - 1);
+  cm->removeRequestsOutOfBatchBounds(8, maxNumOfRequestsInBatch - 1);
+  EXPECT_TRUE(cm->isClientRequestInProcess(8, maxNumOfRequestsInBatch))
+      << "ClientsManager::removeRequestsOutOfBatchBounds removed the request record with the greatest sequence number "
+         "when it should have had fewer than maxNumOfRequestsInBatch request records.";
+  cm->addPendingRequest(8, maxNumOfRequestsInBatch + 1, "correlation ID");
+  cm->addPendingRequest(8, maxNumOfRequestsInBatch + 2, "correlation ID");
+  cm->removeRequestsOutOfBatchBounds(8, maxNumOfRequestsInBatch - 1);
+  EXPECT_TRUE(cm->isClientRequestInProcess(8, maxNumOfRequestsInBatch + 2))
```

Comment:
I have removed this particular check and its associated setup in a recent amendment to my commit for this PR since I went ahead and did so in other cases where the tests were relying on `addPendingRequest` to not enforce limits we think it probably should be enforcing going forward.

I also thought it was odd `removeRequestsOutOfBatchBounds` uses the global system constant `maxNumOfRequestsInBatch` instead of the effective maximum client batch size, and I would be curious to hear if this is found to have been intentional or not.

Reply:
Hey, I spoke with @yontyon about removeRequestsOutOfBatchBounds. It appears that maxNumOfReqsPerClient_ should've been used, but it's best if he confirms it 100%. The implication here is that because the global constant is probably larger than maxNumOfReqsPerClient_, removeRequestsOutOfBatchBounds might not have been able to have any effect. Meaning that fixing this method now is subject to testing.